DOC-221 New Rate-Limiting Article. (#652)
JupiterOne/docs: knowledgeBase/APIs/rate-limiting-j1ql.md

# J1QL Query Rate Limiting

J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query consumes one request token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API. You should incorporate retry with backoff in response to rate-limiting responses from the J1 API.

#### **Account Rate Limit for J1QL Queries**

- Default limit: 120 tokens/120 seconds

> - Enterprise user upgrade: 60 tokens/60 seconds

Reviewer: ```suggestion
- Enterprise limit (available on request): 60 tokens/60 seconds
```

> - Request for approval: 120 tokens/60 seconds

Reviewer: ```suggestion
```
I think this is redundant.

Author: Ok, removing line 9. Thanks!

# J1QL Query Rate Limiting

> J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query uses one token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API.

Reviewer 2: ```suggestion
J1QL queries executed via the [J1 public API](./jupiterone-api.md) are rate-limited separately from [GraphQL API invocations](./rate-limiting.md). Each J1QL query consumes one request token. All queries executed via the J1 public API count towards the rate limit. This count includes user-initiated application queries (such as search and J1 Insights) and queries via the J1 public API.
```

#### **Account Rate Limit for J1QL Queries**

> The default rate of J1QL queries is`120 tokens per 1-minute period`.

Reviewer: ```suggestion
The default rate of J1QL queries is`120 tokens per 2-minute period`.
```

- At any time, there are a maximum of 120 tokens available for J1QL queries.

> - JupiterOne declines queries when the rate exceeds 120 tokens per 1-minute period.

Reviewer: ```suggestion
- JupiterOne declines queries when the rate exceeds 120 tokens per 2-minute period.
```

> - **Average rate allowed:** 2 J1QL queries per second.

Reviewer: ```suggestion
- **Average rate allowed:** 1 J1QL query per second.
```

> - **Average rate allowed:** 2 J1QL queries per second.

Reviewer 2: @benrj is this correct? avg rate allowed is 2 queries / sec?

Reviewer: It should be 1 query per second at the default level.

> - **Configurable:** JupiterOne can adjust this rate limit on your behalf. Contact your J1 administrator.

Reviewer: I'm not sure who is best to make a determination on this point, but I think we want to tie increase or removal of this limit to some kind of business value.   E.g. do we automatically approve all requests from enterprise customers to up to the next level (has been our typical first upgrade)?  

Maybe we could implement something like the following:

**Default limit**: 120 tokens / 120 seconds
**No questions asked upgrade for enterprise customers**: 60 tokens / 60 seconds
**Request for approval**: 120 tokens / 60 seconds
**Not public, select customers only**: no rate limit

And then we'd need to determine what that request for approval process looks like.

@a-u-h-g

Reviewer 2: Agreed with the above proposal. How do you think we should change our current public facing info in sales purchase orders if at all? "30 requests/min = default, 60 requests/min = no question upgrade for enterprise"

Author: @benrj Can we make it a bit more reader friendly this way:

Default limit: 120 tokens / 120 seconds
Enterprise user upgrade: 60 tokens / 60 seconds
Request for approval: 120 tokens / 60 seconds
Select users: no rate limit

So, I'm replacing lines 9-12 with the above?

Reviewer: I would suggest replacing just line 12 with:

```
Enterprise customers are eligible for a rate limit increase up to 120 tokens / 60 seconds, subject to usage analysis and approval.  Contact your TSE to request an increase.

All customers should incorporate retry with backoff in response to rate limiting responses from the J1 API.
```
